### ISSUE TITLE
CA2008 Do not create tasks without passing a TaskScheduler should not appear

### ISSUE THREAD

**Issue by NtFreX** [dotnet/roslyn-analyzers#1907]
#### Analyzer package

[Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

#### Package Version

Example: v2.6.2

#### Diagnostic ID

`CA2008: Do not create tasks without passing a TaskScheduler`
> Do not create tasks unless you are using one of the overloads that takes a TaskScheduler. The default is to schedule on TaskScheduler.Current, which would lead to deadlocks. Either use TaskScheduler.Default to schedule on the thread pool, or explicitly pass TaskScheduler.Current to make your intentions clear.

I could not find a documentation page.

#### Repro steps

```
public class Program
{
  public static async Task Main(string[] args)
  {
    var taskScheduler = TaskScheduler.Default;
    var taskFactory = new TaskFactory(taskScheduler);
    await taskFactory.StartNew(async () => await Task.Delay(1000)).Unwrap();
  }
}
```

#### Expected behavior

No warning is displayed as a `TaskScheduler` has been defined.

#### Actual behavior

Warning is dispalyed.

-----

I've taken a look [here](https://github.com/dotnet/roslyn-analyzers/blob/master/src/Microsoft.NetCore.Analyzers/Core/Tasks/DoNotCreateTasksWithoutPassingATaskScheduler.cs) and in my opinion the code doesn't make sense as `StartNew` on `TaskFactory` has no overload to specify a `TaskScheduler` and its defined when initializing the `TaskFactory`.

**Comment by sharwell** [comment]
My thoughts here:

1. It would be nice for the analyzer to handle this specialized case where a `TaskScheduler` was provided to a custom `TaskFactory` instance, as opposed to obtaining the factory through the `Task.Factory` property. However, this analysis would be relatively expensive and I'm not sure how much benefit it would provide in the end.
2. A scheduler can be provided to `StartNew` via [this overload](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.taskfactory.startnew?view=netframework-4.7.2#System_Threading_Tasks_TaskFactory_StartNew_System_Action_System_Threading_CancellationToken_System_Threading_Tasks_TaskCreationOptions_System_Threading_Tasks_TaskScheduler_).
3. The ideal way to handle this for the case of `TaskScheduler.Default` is to use `Task.Run` instead of `Task.Factory.StartNew`. The call is shorter, the scheduler is consistent, and it avoids the need to call `Unwrap()`.

**Comment by NtFreX** [comment]
@sharwell Thanks for your quick answer.

1. & 2. I'll take a look if I will dig into the code of the analyzer and provide a pullrequest, if I'll just disable the warning or if I use the overload to specify the `TaskScheduler` everytime even if its not needed.
3. As I'm implementing library code where the user should have the possiblity to define the `TaskScheduler` I can't do this.

Most likely I'll use the overload you mentioned.

**Comment by AArnott** [comment]
Amusingly, our version of this analyzer (VSTHRD105) got [the same feedback](https://github.com/Microsoft/vs-threading/issues/370).

**Comment by juliusfriedman** [comment]
This is still a problem with the latest analyzer:

```c#
  var test = Task.Run(async () =>
            {
                await Task.Delay(1000).ConfigureAwait(false);
            });
```

No warning.

Also no warning when you pass the `CancellationToken` and not options and a scheduler.

**Comment by sharwell** [comment]
> This is still a problem with the latest analyzer:

CA2008 does not apply to `Task.Run`. This method uses `TaskScheduler.Default`, as opposed to `TaskFactory.StartNew` that defaults to `TaskScheduler.Current`.
